intake/source/csv.py

```diff
@@ -81,46 +84,115 @@ def _path_column(self):
+            urlpath = [urlpath]
 
-        if self._dataframe is None:
-            self._open_dataset(urlpath)
+        glob_in_path = any("*" in path for path in urlpath)
+        if self.pattern is None and not glob_in_path:
+            self._files = urlpath
+        else:
+            from fsspec.core import get_fs_token_paths
 
-        dtypes = self._dataframe._meta.dtypes.to_dict()
-        dtypes = {n: str(t) for (n, t) in dtypes.items()}
-        return base.Schema(dtype=dtypes, shape=(None, len(dtypes)), npartitions=self._dataframe.npartitions, extra_metadata={})
+            self._files = get_fs_token_paths(urlpath)[2]
+
+        return sorted(self._files)
+
+    def _get_schema(self):
+        if self._schema is not None:
+            return self._schema
+
+        if self._dask_df is not None:
+            dtypes = self._dask_df._meta.dtypes.to_dict()
+            dtypes = {n: str(t) for (n, t) in dtypes.items()}
+            self._schema = base.Schema(dtype=dtypes, shape=(None, len(dtypes)), npartitions=self._dask_df.npartitions, extra_metadata={})
+            return self._schema
+
+        nrows = self._csv_kwargs.get("nrows")
+        self._csv_kwargs["nrows"] = 10
+        df = self._get_partition(0)
+        if nrows is None:
+            del self._csv_kwargs["nrows"]
+        else:
+            self._csv_kwargs["nrows"] = nrows
+
+        dtypes = {col: str(dtype) for col, dtype in df.dtypes.items()}
+        self._schema = base.Schema(dtype=dtypes, shape=(None, len(dtypes)), npartitions=len(self.files()), extra_metadata={})
+
+        return self._schema
 
     def _get_partition(self, i):
-        self._get_schema()
-        return self._dataframe.get_partition(i).compute()
+        if self._dask_df is not None:
+            # FIX: side effects from having different partition
+            # definitions (dask vs. files)?
+            return self._dask_df.get_partition(i).compute()
+
+        url_part = self.files()[i]
+        return self._read_pandas(url_part, i)
+
+    def _read_pandas(self, url_part, i):
+        import pandas as pd
+
+        if self.pattern is None:
+            return pd.read_csv(url_part, storage_options=self._storage_options, **self._csv_kwargs)
+
+        drop_path_column = "include_path_column" not in self._csv_kwargs
+        path_column = self._path_column()
+
+        csv_kwargs = self._csv_kwargs
+        csv_kwargs.pop("include_path_column")
+        df_part = pd.read_csv(url_part, storage_options=self._storage_options, **csv_kwargs)
+
+        from pandas.api.types import CategoricalDtype
+
+        df_part[path_column] = pd.Series(url_part, index=df_part.index, dtype=CategoricalDtype(self.files()))
+        df_part = self._set_pattern_columns(df_part, path_column)
+        if drop_path_column:
+            df_part = df_part.drop([path_column], axis=1)
+
+        return df_part
 
     def read(self):
+        if self._dask_df is not None:
+            return self._dask_df.compute()
```

Comment:
We may revisit this - there is an argument to not use dask here even if to_dask was previously called.

Reply:
Yeah this feels a little squirrelly to me. I think the _registered backend_ idea would resolve this by prescribing a reader.